0009-comit-btsieve-db.adoc

```diff
@@ -0,0 +1,245 @@
+
+==== Document storage
+
+Document-based storage engines are similar to `Key/Value` stores but offer additional functionality usually in the form of "collections".
+Collections are usually used to group together documents of the same kind, i.e. we could have a collection for all swaps using the `rfc003` protocol.
+Within those collections, documents usually have a kind of primary key or index that allows you to retrieve a single document out of the collection.
+
+This is more or less semantically equivalent to the namespacing solution proposed in the `Key/Value` backend.
+In fact, it would be fairly trivial to build our own document-based storage on top of a `Key/Value` stores that implements a particular namespacing algorithm.
+
+For a document-based storage backend, we would have to:
+
+- decide on which collections we want
+- decide how to serialize our structs (JSON, YML, etc)
+
+== Approach
+
+Given the above goals, we need to find an embeddable Rust database engine that supports atomic reads and writes.
+The research will therefore focus on checking the ecosystem and making a recommendation for one of the surveyed libraries.
+
+== Research
+
+Research resulted in the following database engines to be found:
+
+=== seld
+
+Type: `Key/Value` +
+https://crates.io/crates/sled/0.24.1
+
+- + Written in pure Rust
+- + Actively developed
+
+- - alpha state (not yet 1.0, changing binary format between versions)
+
+=== rocksdb
+
+Type: `Key/Value` +
+https://crates.io/crates/rocksdb/0.12.2
+
+- + RocksDB is developed by Facebook
+- + Popular (Parity uses it for substrate f.e.)
+
+=== pickledb
+
+Type: `Key/Value` +
+https://docs.rs/pickledb/0.4.0/pickledb/
+
+- - Not very popular (only 31 stars at the time of writing)
+
+=== unqlite
+
+Type: `Key/Value` & `Document` +
+https://crates.io/crates/unqlite
+
+- + Promises many features (UnQLite is a in-process software library which implements a self-contained, serverless, zero-configuration, transactional NoSQL database engine. UnQLite is a document store database similar to MongoDB, Redis, CouchDB etc. as well a standard Key/Value store similar to BerkeleyDB, LevelDB, etc.)
+
+- - Claims to be actively maintained but not much happening on the repository (could also be due to being considered feature-complete)
+
+=== leveldb
+
+Type: `Key/Value` +
+https://crates.io/crates/leveldb
+
+- + Popular embedded DB written by Google
+
+- - Very low-level interface
+- - Written in C++ and Rust bindings are not complete
+
+There is a re-write of leveldb in Rust https://crates.io/crates/rusty-leveldb[here] but that is not yet complete unfortunately.
+
+=== sqlite
+
+Type: `SQL` +
+https://crates.io/crates/rusqlite
+
+- + actively maintained
+- + quite popular (500 stars)
+- + optionally comes with a bundled version of the `sqlite`-source code to build it in environemnts that don't have it installed
+- + sqlite supports transactions
```

Comment:
What does this mean?

Reply:
Regular transaction with commit and rollback as known from the SQL world.